### ISSUE TITLE
Issues with "Mask" language

### ISSUE THREAD

**Issue by rindeal** [github/linguist#4641]
This is a complaint about the `Mask` entry in `languages.yaml` added in https://github.com/github/linguist/pull/938

## Preliminary Steps

Please confirm you have...
- [x] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
- [x] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
- [x] considered implementing an [override](https://github.com/github/linguist#overrides),
- [x] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).

## Problem Description

The "Mask" (real name "MaskJS") language recognition is currently relying on file extensions alone and as such is claiming all files with `.mask` extension. This extension is generic enough for it to result in much more false positives than actual hits.

My suggestions:

- rename the language from generic "Mask" to specific "MaskJS"
- add some heuristic and don't rely on the file extension alone

OR simply remove "Mask" completely, because it's not a particularly popular markup anyway (just [87 stargazers](https://github.com/atmajs/MaskJS/stargazers) in 6 years).

### URL of the affected repository:
- https://github.com/ukidabek/Dual-Stick-Cover-Shooter/blob/d0665b49e9a9dd8047775e3625602b40b51ad427/Assets/Prefabs/SciFiWarrior/Animators/SciFiWorriorShootMask.mask
- https://github.com/gentoo/gentoo/blob/b37853e23cd9319be6e460905b6ac67400afa535/profiles/base/package.use.mask

### Expected language:

Various languages, eg. Gentoo's [`package.mask` file](https://wiki.gentoo.org/wiki//etc/portage/package.mask), Unity3D Mask assets

### Detected language:

Mask (MaskJS)

**Comment by lildude** [comment]
> The "Mask" (real name "MaskJS") language recognition is currently relying on file extensions alone and as such is claiming all files with `.mask` extension. This extension is generic enough for it to result in much more false positives than actual hits.

It is now, but it wasn't 6 years ago 😉. This affects any language that is introduced after Linguist already has support for an extension it plans to use. In such cases the new language needs to be added to Linguist, which as per the CONTRIBUTING.md would require the addition of a heuristic to help differentiate the two.

In this case no one has submitted a PR for `.mask` to be associated with any other languages.

> My suggestions:
> 
> * rename the language from generic "Mask" to specific "MaskJS"

Seems reasonable to me.

> * add some heuristic and don't rely on the file extension alone

This can only happen when other languages are added as Linguist doesn't have a concept of an unknown language to which we can fallback to.
 
> OR simply remove "Mask" completely, because it's not a particularly popular markup anyway (just [87 stargazers](https://github.com/atmajs/MaskJS/stargazers) in 6 years).

The number of stargazers is not an accurate reflection of the popularity. We base the popularity on the overall usage on GitHub.com. As for deleting it, we don't remove a language once it has been added to Linguist, regardless of how popular or dead it is. We've removed an extension from a language that has multiple extensions on a few rare occasions as the language can still be set by an override, but in this case `.mask` is the _only_ extension associated with this language so we can't remove the extension either.

In summary, you can get the language recognition you wish by adding a new language to Linguist for that extension. Please feel free to review the CONTRIBUTING.md file for details on how to do this and submit a pull request.

**Comment by pchaigno** [comment]
> > My suggestions:
> > 
> > * rename the language from generic "Mask" to specific "MaskJS"
> 
> Seems reasonable to me.

/cc @tenbits who added support for the Mask language in #938. If he's okay with the name change, I can open the pull request for that.

**Comment by rindeal** [comment]
> As for deleting it, we don't remove a language once it has been added to Linguist, regardless of how popular or dead it is. We've removed an extension from a language that has multiple extensions on a few rare occasions as the language can still be set by an override, but in this case `.mask` is the _only_ extension associated with this language so we can't remove the extension either.
> 
> In summary, you can get the language recognition you wish by adding a new language to Linguist for that extension. Please feel free to review the CONTRIBUTING.md file for details on how to do this and submit a pull request.

Let me clarify my point. The `.mask` extension is so generic that any attempt at language recognition based on this extension will inevitably fail, because even if linguist would support all formats that exist today with this extension, there will be two new ones created tomorrow.

You see, MaskJS' addition predates _CONTRIBUTING.md_ by 9 months, i.e. it was added back in the days when no policies existed or were enforced, in fact it was merged without a single comment from <i>linguist</i>s maintainers at that time. Because if the policies that exist today had been applied when MaskJS was being added, it would've been quickly found that there were many more (or rather all) non-MaskJS files with that extension than there were MaskJS ones.

Anyway, I have came up with yet another idea on how to resolve this issue. Apart from adding heuristics to MaskJS language recognition or removing the support altogether, I suggest changing the extension to `.maskjs`, which appears not to be used by any file on GitHub.

**Comment by lildude** [comment]
> Let me clarify my point. The `.mask` extension is so generic that any attempt at language recognition based on this extension will inevitably fail, because even if linguist would support all formats that exist today with this extension, there will be two new ones created tomorrow.

Indeed, and that's no different to any other language that chooses to use a seemly generic or otherwise extension. We'll always be playing catch-up as Linguist is a really simple language detection library which prioritises the extension over any other method. It wouldn't be a trivial exercise to change this either.

> You see, MaskJS' addition predates _CONTRIBUTING.md_ by 9 months, i.e. it was added back in the days when no policies existed or were enforced, in fact it was merged without a single comment from linguists maintainers at that time. Because if the policies that exist today had been applied when MaskJS was being added, it would've been quickly found that there were many more (or rather all) non-MaskJS files with that extension than there were MaskJS ones.

Yes, quite possibly, but we've missed that boat.

> Anyway, I have came up with yet another idea on how to resolve this issue. Apart from adding heuristics to MaskJS language recognition or removing the support altogether, I suggest changing the extension to `.maskjs`, which appears not to be used by any file on GitHub.

That would certainly work, but would require submitting the proposal to the upstream language and then having it accepted and implemented. And then the extension can only be added to Linguist once it reaches our popularity requirements.

I know this isn't ideal, but it's an unfortunate situation we're in due to historical decisions and implementations and we'll continue to be in for quite some time to come.

**Comment by tenbits** [comment]
> The .mask extension is so generic that any attempt at language recognition based on this extension will inevitably fail

@rindeal, looks like you want that files with `mask` extension are of the plain text format and no language should be set behind the extension? Or due to your "Expected language" - you will be happy with anything except the mask syntax? Though Unity3D example could be quite confusing, as here `yaml` syntax is used for the file. And overall, in Unity a file extension often represents not the actual file type, but the entity(_logical_) type in a project: `.asset`, `.meta`, `.anim`, `.mat`, some people also use `.camera`, `.scene`, etc. And all this files are of `yaml` syntax. 


@pchaigno, it seems the announced issue is with the extension itself. Though no other markup syntax, nor language is using `.mask` as its file type. The examples above are logical extension namings (grouping). This can happen to other file extensions as well: `.obj`, `.meta`, `.tool`, `.kit`, `.mod`, `.os` ...

It is up to you and the team to decide how this will proceed. Unfortunately, I can't insist on anything, but I hope, language removal will be not the option, as I and others heavily use the files with the syntax. If you decide to remove/rename the extension for the language, one can still use `.gitattributes` or `vim ft` hints to set the proper syntax.

**Comment by lildude** [comment]
@tenbits we won't be removing the MaskJS language. I do however have an idea on how to address this based on your comment.

@rindeal what prompted you to open this issue? You've covered two different languages with files from two different repos but haven't detailed which you're affected by thus prompting this issue.

**Comment by rindeal** [comment]
I raised this issue, because MaskJS' current matching rules (`.mask` extension) are giving >99.9% false positives including <i>Gentoo</i>s `*.mask` files, which I primarily care about.

Since MaskJS is going nowhere I'd like to propose several partial fixes:

## 1. rename `Mask` to `MaskJS`
This will resolve the confusion about which `*.mask` file is which.

## 2. add `.mask` extension to the list of Unity3D extensions already present in `languages.yaml`
If point 3.2 below won't be accepted, resulting clash with MaskJS would need heuristic based on the 1st line of <i>Unity3D</i>s files which always begins with the same Unity3D-specific string: `%YAML`.

## 3. Solve false positives on <i>Gentoo</i>s `*.mask` files

I have 3 alternatives for this:

### 3.1 add heuristic to MaskJS
This would make sure it doesn't result in so many false positives for <i>Gentoo</i>s `*.mask` files and potentially other similar generic files. The heuristic would need to be complex since I can't seem to see any uniquely identifiable patterns in MaskJS markup.

### 3.2 change MaskJS' extension to `.maskjs`
This would be the simplest and most robust solution, taking care of <i>Gentoo</i>s `*.mask` files as well as of all others, now and in the future.

### 3.3 add a new language - Gentoo's `*.mask` files
These files don't need to be matched by extension, since their whole filename is fixed. The list of the filenames is:
```
package.mask
package.use.mask
package.use.stable.mask
use.mask
use.stable.mask
```

However, since Gentoo uses many more similar files (all sharing roughly the same syntax), this would make unequal conditions for them as some would get recognition, but most not.

**Comment by lildude** [comment]
> including Gentoos `*.mask` files, which I primarily care about.

That's what I was looking and is what I suspected the case would be.  

My idea this morning is a combination of your points and should catch most instances without any major changes:

1. Add `.mask` to Unity3D and mark as generated (they appear generated from the few I've looked at) or implement a heuristic as looks easier to detect, with a fallback to MaskJS.
2. Add the Gentoo mask files as defined filenames to the Text language.

This should a majority of files. The only exception will be when users get creative with their filenames, but we have that problem now anyway 😄
